### PR TITLE
OHRM5X-2001: Claim event filter form

### PR DESCRIPTION
**Pull request by DulsaraNethmin**

### CODE REVIEW

File: installer/Migration/V5_4_0/Migration.php [hunk 1]
```diff
@@ -148,6 +153,56 @@ public function up(): void
...
+                    ->setMaxResults($batchSize)
+                    ->executeQuery();
+            foreach ($result->fetchAllAssociative() as $row) {
+                $this->createQueryBuilder()
+                        ->update($table)
+                        ->set('status', ':status')
+                        ->where('id = :id')
+                        ->setParameter('id', $row['id'])
+                        ->setParameter('status', $row['status']=='on' ? 1 : 0)
+                        ->executeStatement();
```
Comment (ManulMax) [inline installer/Migration/V5_4_0/Migration.php:189]: need to check this twise

File: installer/Migration/V5_4_0/Migration.php [hunk 2]
```diff
@@ -148,6 +153,56 @@ public function up(): void
...
+    {
+        $count = $this->createQueryBuilder()
+            ->select("COUNT($tableName.id)")
+            ->from($tableName)
+            ->executeQuery()
+            ->fetchOne();
+        Logger::getLogger()->info("`$tableName` record count: $count");
+        return $count;
+    }
+
```
Comment (ManulMax) [inline installer/Migration/V5_4_0/Migration.php:205]: can remove this and implement according to the `changeClaimExpenseTypeTableStatusToBoolean` after rebasing

File: installer/Migration/V5_4_0/Migration.php (hunk 1 above)
Comment (ManulMax) [reply installer/Migration/V5_4_0/Migration.php:189]: implement according to the changeClaimExpenseTypeTableStatusToBoolean after rebasing

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+              name: 'pencil-fill',
+            },
+          },
+        },
+      },
+    ],
+    checkedItems: [],
+    ClaimEventStatuses: [
+      {id: 1, label: 'Enabled'},
+      {id: 0, label: 'Disabled'},
```
Comment (RajithaKumara) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:198]: i18n

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 4]
```diff
@@ -79,7 +102,36 @@ public function setClaimEvent(ClaimEvent $claimEvent)
     }
 
     /**
-     * @return EndpointResult
+     * @OA\Get(
+     *     path="/api/v2/claim/events",
+     *     tags={"Claim/Claim Events"},
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:107]:
```suggestion
     *     tags={"Claim/Events"},
```
@ManulMax what do you think?

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: installer/Migration/V5_4_0/lang-string/claim.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+langStrings:
+  -
+    value: 'Claim'
+    unitId: claim
+  -
+    value: 'Events'
+    unitId: events
```
Comment (Super-Chama) [inline installer/Migration/V5_4_0/lang-string/claim.yaml:7]: add new line in the end

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+      checkedItems: [],
+      ClaimEventStatuses: [
+        {id: 1, label: this.$t('Active')},
+        {id: 0, label: this.$t('Inactive')},
+      ],
+    };
+  },
+
+  computed: {
+    items_new() {
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:217]: you can use a `normalizer` function as a parameter for `usePaginate` function. `normalizer` is a function that can transform raw data retrived via API before getting output.

https://github.com/orangehrm/orangehrm/blob/main/src/client/src/orangehrmAdminPlugin/pages/systemUser/SystemUser.vue#L200

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        },
+        {
+          name: 'actions',
+          title: this.$t('general.actions'),
+          slot: 'action',
+          style: {flex: 1},
+          cellType: 'oxd-table-cell-actions',
+          cellConfig: {
+            delete: {
+              onClick: () => {
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:189]: `onClick: this.onClickDelete`

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                this.onClickDelete();
+              },
+              component: 'oxd-icon-button',
+              props: {
+                name: 'trash',
+              },
+            },
+            edit: {
+              onClick: () => {
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:198]: implement a dummy method with body as `//TODO`

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+  methods: {
+    async resetDataTable() {
+      this.checkedItems = [];
+      await this.execQuery();
+    },
+    async filterItems() {
+      await this.execQuery();
+    },
+    onClickReset() {
+      this.filter2 = null;
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:238]: `filter2` seems unused

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+      this.filters = {...defaultFilters};
+      this.filterItems();
+    },
+    onClickAdd() {
+      navigate('/claim/events/save');
+    },
+    onClickDelete() {
+      this.$refs.deleteDialog.showDialog().then((confirmation) => {
+        if (confirmation === 'ok') {
+          //this.deleteItems([item.id]);
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:248]: not implemented?

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        }
+      });
+    },
+  },
+};
+</script>
+
+<style></style>
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:256]: remove empty block

File: src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+ -->
+
+<template>
+  <oxd-table-filter filter-title="Events">
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/pages/ClaimEvent.vue:22]: Events should be i18n string

File: src/client/src/orangehrmClaimPlugin/index.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+import ClaimEvent from '@/orangehrmClaimPlugin/pages/ClaimEvent.vue';
+import SaveClaimEvent from '@/orangehrmClaimPlugin/pages/SaveClaimEvent.vue';
```
Comment (Super-Chama) [inline src/client/src/orangehrmClaimPlugin/index.ts:20]: this file is missing?

**Review by Super-Chama — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php (hunk 4 above)
Comment (ManulMax) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:107]: tags={"Claim/Events"}, is better

File: src/client/src/orangehrmClaimPlugin/index.ts (hunk 13 above)
Comment (DulsaraNethmin) [reply src/client/src/orangehrmClaimPlugin/index.ts:20]: i didn't commit it

File: src/plugins/orangehrmClaimPlugin/test/Api/ClaimExpenseTypeAPITest.php [hunk 14]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/test/Api/ClaimExpenseTypeAPITest.php:2]: remove this

File: src/plugins/orangehrmClaimPlugin/test/Api/ClaimRequestAPITest.php [hunk 15]
```diff
@@ -24,10 +24,6 @@
 use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
 use OrangeHRM\Tests\Util\Integration\TestCaseParams;
 
-/**
- * @group Claim
- * @group APIv2
- */
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/test/Api/ClaimRequestAPITest.php:30]: don't remove these codes

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 16]
```diff
@@ -215,7 +294,48 @@ public function getValidationRuleForGetOne(): ParamRuleCollection
     }
 
     /**
-     * @return EndpointResult
+     * @OA\Put(
+     *     path="/api/v2/claim/events/{id}",
+     *     tags={"Claim/Claim Events"},
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:299]: change this tags according to the feedback

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 17]
```diff
@@ -190,8 +249,28 @@ public function getValidationRuleForDelete(): ParamRuleCollection
     }
 
     /**
-     * @return EndpointResult
-     * @throws RecordNotFoundException
+     * @OA\Get(
+     *     path="/api/v2/claim/events/{id}",
+     *     tags={"Claim/Claim Events"},
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:254]: tags={"Claim/Events"},

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 18]
```diff
@@ -56,6 +56,29 @@ class ClaimEventAPI extends Endpoint implements CrudEndpoint
     public const NAME_MAX_LENGTH = 100;
 
     /**
+     * @OA\Post(
+     *     path="/api/v2/claim/events",
+     *     tags={"Claim/Claim Events"},
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:61]: tags={"Claim/Events"},

**Review by ManulMax — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 19]
```diff
@@ -215,7 +294,48 @@ public function getValidationRuleForGetOne(): ParamRuleCollection
...
+     *                 type="string"
+     *             ),
+     *             @OA\Property(
+     *                 property="status",
+     *                 type="boolean"
+     *             ),
+     *             @OA\Property(
+     *                 property="isDeleted",
+     *                 type="boolean"
+     *             )
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:322]:
```suggestion
```

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 20]
```diff
@@ -215,7 +294,48 @@ public function getValidationRuleForGetOne(): ParamRuleCollection
...
+     *                 property="data",
+     *                 ref="#/components/schemas/Claim-ClaimEventModel"
+     *             ),
+     *             @OA\Property(property="meta", type="object")
+     *         )
+     *     ),
+     *     @OA\Response(response="404", ref="#/components/responses/RecordNotFound")
+     * )
+     * @inheritDoc
      * @throws RecordNotFoundException
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:339]:
```suggestion
```

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimEventModel.php [hunk 21]
```diff
@@ -24,6 +24,28 @@
...
+ *         property="description",
+ *         type="string",
+ *     ),
+ *     @OA\Property(
+ *         property="status",
+ *         type="boolean",
+ *     )
+ * )
+ */
 class ClaimEventModel implements Normalizable
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimEventModel.php:49]: expose `deleted` parameter as well

Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/Model/ClaimEventModel.php:49]: ref: src/plugins/orangehrmLeavePlugin/Api/Model/LeaveTypeModel.php

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 22]
```diff
@@ -79,7 +102,36 @@ public function setClaimEvent(ClaimEvent $claimEvent)
...
+     *     tags={"Claim/Events"},
+     *     @OA\Parameter(
+     *         name="sortField",
+     *         in="query",
+     *         required=false,
+     *         @OA\Schema(type="string", enum=ClaimEventSearchFilterParams::ALLOWED_SORT_FIELDS)
+     *     ),
+     *     @OA\Parameter(ref="#/components/parameters/sortOrder"),
+     *     @OA\Parameter(ref="#/components/parameters/limit"),
+     *     @OA\Parameter(ref="#/components/parameters/offset"),
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:116]: `eventId` test scenarios

1. zero
2. existing id
3. non-existing id
4. try deleted id

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimEventModel.php (hunk 21 above)
Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/Model/ClaimEventModel.php:49]: No need to handle this

File: installer/Migration/V5_4_0/lang-string/claim.yaml (hunk 5 above)
Comment (RajithaKumara) [reply installer/Migration/V5_4_0/lang-string/claim.yaml:7]: this feedback not addressed

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 23]
```diff
@@ -50,12 +49,36 @@ class ClaimEventAPI extends Endpoint implements CrudEndpoint
     public const PARAMETER_NAME = 'name';
     public const PARAMETER_DESCRIPTION = 'description';
     public const PARAMETER_ID = 'id';
+    public const PARAMETER_CLAIMID = 'claimId';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:52]:
```suggestion
    public const PARAMETER_EVENT_ID = 'eventId';
```
refactor this

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 24]
```diff
@@ -50,12 +49,36 @@ class ClaimEventAPI extends Endpoint implements CrudEndpoint
     public const PARAMETER_NAME = 'name';
     public const PARAMETER_DESCRIPTION = 'description';
     public const PARAMETER_ID = 'id';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:51]: @ManulMax can use from common parms

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php (hunk 22 above)
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:116]: not addressed

File: src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php [hunk 25]
```diff
@@ -114,6 +166,14 @@ public function getValidationRuleForGetAll(): ParamRuleCollection
                     )
                 )
             ),
+            $this->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::PARAMETER_CLAIMID,
+                    new Rule(
+                        Rules::STRING_TYPE
+                    )
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimEventAPI.php:174]:
```suggestion
                    new Rule(Rules::POSITIVE)
```

File: src/plugins/orangehrmClaimPlugin/Dto/ClaimEventSearchFilterParams.php [hunk 26]
```diff
@@ -23,7 +23,7 @@
 
 class ClaimEventSearchFilterParams extends FilterParams
 {
-    public const ALLOWED_SORT_FIELDS = ['claimEvent.name','claimEvent.status'];
+    public const ALLOWED_SORT_FIELDS = ['claimEvent.name','claimEvent.status','claimEvent.id'];
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dto/ClaimEventSearchFilterParams.php:26]:
```suggestion
    public const ALLOWED_SORT_FIELDS = ['claimEvent.name','claimEvent.status'];
```
remove `claimEvent.id`

File: src/plugins/orangehrmClaimPlugin/test/fixtures/ClaimEvent.yml [hunk 27]
```diff
@@ -14,7 +14,8 @@ User:
   - { "id": 3, "user_role_id": 2, "emp_number": 2, "user_name": "Ashley","user_password": "Ashley","deleted": "0","status": "1","date_entered": "2022-09-15 16:47:14","created_by": "1" }
 
 ClaimEvent:
-  0: { id: 1, name: "event1", description: "event1 desc", status: true }
-  1: { id: 2, name: "event2", description: "event2 desc", status: true }
-  2: { id: 3, name: "event3", description: "event3 desc", status: false }
-  3: { id: 4, name: "event4", description: "event4 desc", status: false }
+  0: { id: 1, name: "event1", description: "event1 desc", status: true, is_deleted: false}
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/test/fixtures/ClaimEvent.yml:17]: prefer if you can define meaningful data

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Pull request merged**